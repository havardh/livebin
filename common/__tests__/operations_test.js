import {
  insert,
  del,
  retain,
  generate,
  simplify,
  transform
} from "../operations";

describe("Operations", () => {
  describe("insert", () => {
    test("that it create a message", () => {
      expect(insert("test")).toMatchSnapshot();
    });
  });

  describe("del", () => {
    test("that it create a message", () => {
      expect(del("test")).toMatchSnapshot();
    });
  });

  describe("retain", () => {
    test("that it create a message", () => {
      expect(retain(2)).toMatchSnapshot();
    });
  });

  describe("generate", () => {
    test("that it throws if zero characters are changed", () => {
      expect(
        () => generate("123", "123")
      ).toThrowErrorMatchingSnapshot();
    });

    describe("delete", () => {
      test("first", () => {
        expect(generate("123", "23")).toMatchSnapshot();
      });

      test("inside", () => {
        expect(generate("123", "13")).toMatchSnapshot();
      });

      test("last", () => {
        expect(generate("123", "12")).toMatchSnapshot();
      });

      test("same character", () => {
        expect(generate("111", "11")).toMatchSnapshot();
      });
    });

    describe("add", () => {
      test("before", () => {
        expect(generate("23", "123")).toMatchSnapshot();
      });

      test("inside", () => {
        expect(generate("13", "123")).toMatchSnapshot();
      });

      test("after", () => {
        expect(generate("12", "123")).toMatchSnapshot();
      });

      test("same character", () => {
        expect(generate("11", "111")).toMatchSnapshot();
      });
    })
  });

  describe("simplify", () => {
    test("retain", () => {
      const actual = simplify([retain(1), retain(1)]);
      const expected = [retain(2)];

      expect(actual).toEqual(expected)
    });

    test("insert", () => {
      const actual = simplify([insert("a"), insert("b")]);
      const expected = [insert("ab")];

      expect(actual).toEqual(expected)
    });

    test("del", () => {
      const actual = simplify([del("a"), del("b")]);
      const expected = [del("ab")];

      expect(actual).toEqual(expected)
    });
  });

  describe.only("transform", () => {
    const server = [retain(2), insert("a")];
    const client = [retain(2), insert("b")];

    const { left, right } = transform(server, client);

    expect(left).toEqual([
      retain(2),
      insert("a"),
      retain(1)
    ]);
    expect(right).toEqual([
      retain(3),
      insert("b")
    ]);
  });
});
