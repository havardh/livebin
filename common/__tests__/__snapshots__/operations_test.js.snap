// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Operations del that it create a message 1`] = `
Object {
  "characters": "test",
  "type": "delete",
}
`;

exports[`Operations generate add after 1`] = `
Array [
  Object {
    "n": 2,
    "type": "retain",
  },
  Object {
    "characters": "3",
    "type": "insert",
  },
]
`;

exports[`Operations generate add before 1`] = `
Array [
  Object {
    "characters": "1",
    "type": "insert",
  },
  Object {
    "n": 2,
    "type": "retain",
  },
]
`;

exports[`Operations generate add inside 1`] = `
Array [
  Object {
    "n": 1,
    "type": "retain",
  },
  Object {
    "characters": "2",
    "type": "insert",
  },
  Object {
    "n": 1,
    "type": "retain",
  },
]
`;

exports[`Operations generate add same character 1`] = `
Array [
  Object {
    "n": 2,
    "type": "retain",
  },
  Object {
    "characters": "1",
    "type": "insert",
  },
]
`;

exports[`Operations generate delete first 1`] = `
Array [
  Object {
    "characters": "1",
    "type": "delete",
  },
  Object {
    "n": 2,
    "type": "retain",
  },
]
`;

exports[`Operations generate delete inside 1`] = `
Array [
  Object {
    "n": 1,
    "type": "retain",
  },
  Object {
    "characters": "2",
    "type": "delete",
  },
  Object {
    "n": 1,
    "type": "retain",
  },
]
`;

exports[`Operations generate delete last 1`] = `
Array [
  Object {
    "n": 2,
    "type": "retain",
  },
  Object {
    "characters": "3",
    "type": "delete",
  },
]
`;

exports[`Operations generate delete same character 1`] = `
Array [
  Object {
    "n": 2,
    "type": "retain",
  },
  Object {
    "characters": "1",
    "type": "delete",
  },
]
`;

exports[`Operations generate that it throws if zero characters are changed 1`] = `"not a one character change"`;

exports[`Operations insert that it create a message 1`] = `
Object {
  "characters": "test",
  "type": "insert",
}
`;

exports[`Operations retain that it create a message 1`] = `
Object {
  "n": 2,
  "type": "retain",
}
`;
